File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
 
class Solution {
    int path(TreeNode* root, int& res){
        if(!root)   return 0;
        int left=max(0, path(root->left, res));
        int right=max(0, path(root->right, res));
        res=max(res, left+right+root->val);
        return root->val+max(left, right);
    }
public:
    int maxPathSum(TreeNode* root) {
        int res=INT_MIN;
        path(root, res);
        return res;
    }
};

File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def maxSum(self, root, maxi):
        if not root:
            return 0
        
        left = max(0, self.maxSum(root.left, maxi))
        right = max(0, self.maxSum(root.right, maxi))

        maxi[0] = max(maxi[0], left + right + root.val)

        return max(left, right) + root.val

    def maxPathSum(self, root):
        """
        :type root: Optional[TreeNode]
        :rtype: int
        """
        res = [float('-inf')] 
        self.maxSum(root, res)
        return res[0]